### PR TITLE
WIP Work towards using a vanilla version of ASM [ci: last-only]

### PR DESCRIPTION
**Pull request by retronym**
[Instructions](https://github.com/retronym/asm/pull/7) on how to build the minimally modified asm.

References: https://github.com/scala/scala-asm/issues/9

### CODE REVIEW

File: src/compiler/scala/tools/nsc/backend/jvm/MethodNode1.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+//            }
+//        };
+    }
+
+    @Override
+    protected LabelNode getLabelNode(Label label) {
+        if (!(label.info instanceof LabelNode)) {
+            label.info = new LabelNode1(label);
+        }
+        return (LabelNode) label.info;
```
Comment (retronym) [inline src/compiler/scala/tools/nsc/backend/jvm/MethodNode1.java:37]: I tried pushing this into ASM itself, but doing so breaks a [unit test](https://github.com/retronym/asm/blob/2152b6abc5a49e6bdc2cfd00a23ceac064317c55/asm-tree/src/test/java/org/objectweb/asm/tree/ClassNodeTest.java#L360-L374) which appears to be a legitimate use case the only works under the status quo.

The failure mode for that test is an infinite loop iterating an instruction list. (or, a fail fast from our other customization that adds an assertion in `InsnList.add` to check that the given instruction is not already in another instruction list.

File: src/compiler/scala/tools/nsc/backend/jvm/LabelNode1.java [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+public class LabelNode1 extends LabelNode {
+    public LabelNode1() {
+    }
+
+    public LabelNode1(Label label) {
+        super(label);
+    }
+
+    public int flags;
+}
```
Comment (retronym) [inline src/compiler/scala/tools/nsc/backend/jvm/LabelNode1.java:15]: The ASM maintainers said they didn't want to make `Label.flags` public. I think we can achieve the same result by using this subclass of `LabelNode` with a spot for some user-data.

**Comment by mkurz** [comment]
@retronym Why are you closing this pr?

**Comment by retronym** [comment]
I've rolled most of the changes into #6733 (which I'm still working on).

Once ASM 6.3 is released, we can do the last step, which is to depend on that instead of scala-asm. We have to decide whether we shade the binary. Not shading buys us user-upgradability of ASM but risks conflicts with different versions of ASM required by other libraries on the classpath.

**Pull request closed without merging**